client/src/virtual_layer.rs

```diff
@@ -207,148 +236,169 @@ impl TcpLayer {
+            log::debug!(
+                "[VirtualTcp] Incoming message from new Node [{}]. Adding connection.",
+                node.id
+            );
+
             self.add_virt_node(node).await.ok();
         }
 
         self.net.receive(payload.into_vec());
         self.net.poll();
     }
 
-    fn spawn_ingress_router(&self) -> anyhow::Result<()> {
+    pub async fn shutdown(&self) {
+        let mut state = self.state.write().await;
+
+        for (_, mut sender) in state.forward_senders.drain() {
+            sender.close().await.ok();
+        }
+    }
+
+    async fn spawn_ingress_router(&self) -> anyhow::Result<()> {
         let ingress_rx = self
             .net
             .ingress_receiver()
             .ok_or_else(|| anyhow::anyhow!("Ingress traffic router already spawned"))?;
 
-        let myself = self.clone();
-        tokio::task::spawn_local(ingress_rx.for_each(move |event| {
-            let myself = myself.clone();
-            async move {
-                let (desc, payload) = match event {
-                    IngressEvent::InboundConnection { desc } => {
-                        log::trace!(
-                            "[{}] ingress router: new connection from {:?} to {:?} ",
-                            myself.net_id(),
-                            desc.remote,
-                            desc.local,
-                        );
-                        return;
-                    }
-                    IngressEvent::Disconnected { desc } => {
-                        log::trace!(
-                            "[{}] ingress router: ({}) {:?} disconnected from {:?}",
-                            myself.net_id(),
-                            desc.protocol,
-                            desc.remote,
-                            desc.local,
-                        );
-                        return;
-                    }
-                    IngressEvent::Packet { desc, payload, .. } => (desc, payload),
-                };
-
-                if desc.protocol != Protocol::Tcp {
-                    log::trace!(
-                        "[{}] ingress router: dropping {} payload",
-                        myself.net_id(),
-                        desc.protocol
-                    );
-                    return;
-                }
+        tokio::task::spawn_local(self.clone().ingress_router(ingress_rx));
+        Ok(())
+    }
+
+    async fn spawn_egress_router(&self) -> anyhow::Result<()> {
+        let egress_rx = self
+            .net
+            .egress_receiver()
+            .ok_or_else(|| anyhow::anyhow!("Egress traffic router already spawned"))?;
 
-                let remote_address = match desc.remote {
-                    SocketEndpoint::Ip(endpoint) => endpoint.addr,
-                    _ => {
+        tokio::task::spawn_local(self.clone().egress_router(egress_rx));
+        Ok(())
+    }
+
+    async fn ingress_router(self, ingress_rx: UnboundedReceiver<IngressEvent>) {
+        ingress_rx
```

Comment:
Yes but this PR is improvement, because I split a few functions here.
I'm not touching this in this PR since, this function was like this before, I only moved some code